### DISCUSSION TITLE
Difficulties in rediscovering Von Neumann entropy from eigenvalue data

### DISCUSSION THREAD

**Discussion by AndrewValentini** [MilesCranmer/PySR#685] [Q&A]
The Von Neumann equation states that the entropy of a density matrix $\rho$ is of the form:

$$S(\rho) = -\sum_i \lambda_i \ln \lambda_i$$
where $\{\lambda_i\}$ are the eigenvalues of $\rho$. 

I have 10,000+ rows of data for the eigenvalues for density matrices as they evolve over time and I have 10,000+ rows of data for the entropy of the density matrices. I have checked that the eigenvalue data matches the correct entropy value when the Von Neumann equation is applied to the eigenvalues. I am trying to use PySR to rediscover the Von Neumann entropy equation using the eigenvalue data before I try applying the package to more complicated scenarios in quantum information. 

I can't seem to get PySR to find the correct expression. I'm including a screenshot of its results when working with my data. In this, "l1" and "l2" are the two eigenvalues of the density matrices. My density matrices only ever have two eigenvalues. 

<img width="780" alt="PySR_HOF" src="https://github.com/user-attachments/assets/f61b42f8-530a-4f62-bd17-60b28ecc89da">

... I was assuming that this would have been an easy problem for PySR to solve given that I just have two columns of data and the correct expression just a sum with some natural logs attached. I'm completely new to using PySR so I am assuming that I just don't understand how to work with the package yet. This is the code that I'm using to generate the model, which I've left almost completely unedited from the example given on the repo's README file: 

<img width="1073" alt="PySR_code" src="https://github.com/user-attachments/assets/332b2fa5-d1e2-47ed-86c5-2c16cbd3080f">

Should I be using different operators? Is the number of iterations too small? Is it problematic to try discovering an expression that only depends on two variables? Is it a known issue that PySR struggles to predict expressions involving natural logs and sums?

Any guidance would be greatly appreciated.

**Comment by MilesCranmer** [comment]
Hm, yeah it should be pretty easy for it to find that. Are there any data preprocessing steps that could be generating invalid results for `y`?

Out of curiosity, why not give it `+` too? It is needed for some of the simplification operators to work.

Also since it’s so simple you likely don’t need that many datapoints. Even 100 should be okay.

**Reply by AndrewValentini** [reply]
Wow, thanks for the quick reply. 

Yeah, I'll have to go back and make sure that the entropy data is being collected correctly. I added the addition operator and lowered the sample size down to 100 but wasn't able to find the correct expression. These are the results from that run:

<img width="900" alt="PySR_HOF2" src="https://github.com/user-attachments/assets/5fb93f2f-bd04-4eef-9363-c3758b3dfd2c">

The issue must just be on my end. Thank you for confirming that I'm at least using the package correctly. I would be willing to send the data to you if you're interested in seeing it.

**Reply by AndrewValentini** [reply]
Is there a way to force PySR to only generate equations with just the variables it's given? Maybe that would result in the correct expression pretty quickly since there are only so many ways you can combine addition, subtraction, multiplication, and natural log operators with the two variables provided...

**Reply by AndrewValentini** [reply]
My problems were the result of some errors in saving data. I'm able to discover the Von Neumann entropy expression now. Thanks a ton. Is there any way of forcing PySR to only generate expressions with the defined variables you give it, though?

**Reply by MilesCranmer** [reply]
Yes you can do this with `complexity_of_constants=100` so it will avoid fitting constants

**Reply by MilesCranmer** [reply]
You should check out the docs, maybe https://astroautomata.com/PySR/api/ and https://astroautomata.com/PySR/tuning/
